kratos/utilities/properties_accessor.h

```diff
@@ -0,0 +1,327 @@
+//
+
+#if !defined(KRATOS_PROPERTIES_ACCESSOR_H)
+#define  KRATOS_PROPERTIES_ACCESSOR_H
+
+// System includes
+#include <map>
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "containers/data_value_container.h"
+#include "geometries/geometry.h"
+#include "includes/node.h"
+#include "includes/process_info.h"
+#include "includes/properties.h"
+
+
+namespace Kratos
+{
+///@addtogroup KratosCore
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+*/
+class PropertiesAccessor : public IndexedObject
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of PropertiesAccessor
+    KRATOS_CLASS_POINTER_DEFINITION(PropertiesAccessor);
+
+    using BaseType = IndexedObject;
+
+    using NodeType = Node<3>;
+    using GeometryType = Geometry<NodeType>;
+
+    using IndexType = std::size_t;
+    using SizeType = std::size_t;
+
+    using DoubleVariableType = Variable<double>;
+    using ComponentVariableType = VariableComponent< VectorComponentAdaptor<array_1d<double, 3> > >;
+    using Array3VariableType = Variable<array_1d<double, 3>>;
+
+    using DoubleFunction = std::function<double(const Properties::Pointer&,
+                                                const GeometryType::Pointer&,
+                                                const DataValueContainer&,
+                                                const ProcessInfo&,
+                                                const Vector&,
+                                                const int)>;
+
+    using DoubleConfiguration = std::map<DoubleVariableType, DoubleFunction>;
+
+    // std::function<void(int)>
+
+    // using std::unordered_map<DoubleVariableType>
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    PropertiesAccessor(IndexType NewId = 0) : BaseType(NewId) {}
+
+    /// Destructor.
+    virtual ~PropertiesAccessor() {}
+
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    template<class TVariableType>
+    typename TVariableType::Type& GetValue(const TVariableType& rV,
+                                           const Properties::Pointer& rpProperties,
+                                           const GeometryType::Pointer& rpGeometry,
+                                           const DataValueContainer& rDataContainer,
+                                           const ProcessInfo& rCurrentProcessInfo,
+                                           const Vector& rShapeFunctionValues,
+                                           const int GaussPointIndex=0)
+    {
+        // rpProperties->GetValue(rV);
+    }
+
+
+    double GetValue(const DoubleVariableType& rV,
+                     const Properties::Pointer& rpProperties,
```

Comment:
i would pass Properties when we construct the accessor

Reply:
I would also prefer this, but I don't know how to do it bcs of circular dependencies:
the accessor includes properties, whereas the properties would have to include the accessor.
Forward declaration does not work here I think bcs one would need the actual type

Maybe if I split it into header/souce...
Will give it another try